---
title: "Caenorhabditis spp. Gene Annotations"
author: "Astra Bryant"
date: "2023-12-12"
output: pdf_document
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
The goal of this file is to import and clean gene annotation information for all genes from the following *Caenorhabditis*  species: *C. elegans*, *C. briggsae*, *C. remanei*, *C. japonica*, and *C. brenneri*.

Annotation information includes:
  * Gene name
  * WormBase gene name/IDs
  * WormBase description
  * InterPro description
  * Homologs/Percent homology with other *Caenorhabditis* species  

```{r setup, echo = FALSE, results = FALSE, include = FALSE}
if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!require("tximport", quietly = TRUE)) BiocManager::install("tximport", ask = FALSE)
if (!require("ensembldb", quietly = TRUE)) BiocManager::install("ensembldb", ask = FALSE)
if (!require("biomaRt", quietly = TRUE)) BiocManager::install("biomaRt", ask = FALSE)
if (!require("pacman", quietly = TRUE)) install.packages("pacman")
pacman::p_load("tidyverse","data.table", "magrittr","ensembldb", "biomaRt", "knitr")

# Check for presence of output folder, generate if it doesn't exist
output.path <- "./Outputs"
if (!dir.exists(output.path)){
  dir.create(output.path)
}

app.path <-"../Data"
www.path <-"../www"
```

```{r geneAnnotation}
# Introduction to this chunk -----------
# This chunk imports gene annotation information using WBPS BiomaRT
# It will generate and save a table.

# Get gene information from BiomaRT and filter -----
species_list <- tibble(species = c('caelegprjna13758', 'cabrigprjna10731', 'cabrenprjna20035', 'caremaprjna53967', 'cajapoprjna12591'))

for (x in species_list$species) {
  comparison.attributes <- dplyr::filter(species_list, species!= x ) %>%
  dplyr::mutate(geneID = paste0(species, "_gene"), geneName = paste0(species, "_gene_name"), percID = paste0(species, "_homolog_perc_id"))

annotations <- getBM(attributes=c('wbps_gene_id', 
                                'external_gene_id',
                                'wormbase_gseq',
                                'description',
                                comparison.attributes$geneID[[1]],
                                comparison.attributes$geneName[[1]],
                                comparison.attributes$percID[[1]],
                                comparison.attributes$geneID[[2]],
                                comparison.attributes$geneName[[2]],
                                comparison.attributes$percID[[2]],
                                comparison.attributes$geneID[[3]],
                                comparison.attributes$geneName[[3]],
                                comparison.attributes$percID[[3]],
                                comparison.attributes$geneID[[4]],
                                comparison.attributes$geneName[[4]],
                                comparison.attributes$percID[[4]]
                                ),
                     # grab the annotations from WormBase ParaSite
                     mart = useMart(biomart="parasite_mart", 
                                    dataset = "wbps_gene", 
                                    host="https://parasite.wormbase.org", 
                                    port = 443),
                     filters = 'species_id_1010',
                     useCache = T,
                     value = x) %>%
  as_tibble(.name_repair = "unique") %>%
  tidyr::unite("GS4_homologID", 15, 14, sep = " | ", remove = T) %>%
  tidyr::unite("GS3_homologID", 12, 11, sep = " | ", remove = T) %>%
  tidyr::unite("GS2_homologID", 9, 8, sep = " | ", remove = T) %>%
  tidyr::unite("GS1_homologID", 6, 5, sep = " | ", remove = T) %>%
  dplyr::rename(geneID = "Gene stable ID",
                geneName = "Gene name",
                WormBaseID = "WormBase Sequence Name",
                Description = "Gene description",
                GS1_percent_homology = "% identity...7",
                GS2_percent_homology = "% identity...10",
                GS3_percent_homology = "% identity...13",
                GS4_percent_homology = "% identity...16",
  ) %>%
  dplyr::group_by(geneID)

# Gather then clean interpro description, removing duplications and collapsing to one line
interpro_terms <- getBM(attributes=c('wbps_gene_id', 'interpro_description'),
                     # grab the annotations from WormBase ParaSite
                     mart = useMart(biomart="parasite_mart", 
                                    dataset = "wbps_gene", 
                                    host="https://parasite.wormbase.org", 
                                    port = 443),
                     filters = 'species_id_1010',
                     useCache = T,
                     value = x) %>%
  dplyr::rename(geneID = "wbps_gene_id",
                InterPro = "interpro_description") %>%
  group_by(geneID) %>%
  dplyr::distinct(InterPro, .keep_all = TRUE) %>%
  dplyr::summarise(InterPro = paste(InterPro, collapse = ', '))

annotations <-left_join(annotations, interpro_terms) %>%
  relocate(InterPro, .after = Description)

# Remove source code to shorten the description
annotations$Description <- annotations$Description %>%
  str_replace_all(string = ., 
                  pattern = "  \\[Source:.*\\]", 
                  replacement = "")

# Replace empty string values with NAs
annotations[annotations == "NA | NA"]<-NA
annotations[annotations == " | "]<-NA

# Remove any duplicates in the possible homolog matches. 
# Select based on highest % homology.
# Give fake value here to make sure genes 
# without homologs aren't filtered out
annotations$GS1_percent_homology[
  is.na(annotations$GS1_percent_homology)] <- 1000 
annotations$GS2_percent_homology[
  is.na(annotations$GS2_percent_homology)] <- 1000
annotations$GS3_percent_homology[
  is.na(annotations$GS3_percent_homology)] <- 1000
annotations$GS4_percent_homology[
  is.na(annotations$GS4_percent_homology)] <- 1000

annotations <-annotations %>%
  slice_max(n = 1, order_by = GS1_percent_homology, 
            with_ties = FALSE) %>%
  slice_max(n = 1, order_by = GS2_percent_homology, 
            with_ties = FALSE) %>%
  slice_max(n = 1, order_by = GS3_percent_homology, 
            with_ties = FALSE) %>%
  slice_max(n = 1, order_by = GS4_percent_homology, 
            with_ties = FALSE)

annotations$GS1_percent_homology[
  annotations$GS1_percent_homology == 1000] <- NA
annotations$GS2_percent_homology[
  annotations$GS2_percent_homology == 1000]<- NA
annotations$GS3_percent_homology[
  annotations$GS3_percent_homology == 1000]<- NA
annotations$GS4_percent_homology[
  annotations$GS4_percent_homology == 1000]<- NA


# Data cleaning, required because a small number of genes do not appear in the BioMaRT database. Their information needs to be added.
# How do we know which genes? For *C. elegans* there are four that appear in the modENCODE RNA-seq dataset that do not have corresponding annotations. We are adding those. 
if (x == "caelegprjna13758"){
  supplement <- tibble (geneID = c("WBGene00306134", "WBGene00306133", "WBGene00306132", "WBGene00306131"),
                        geneName = c("srlf-8", "azyx-1", "F54D10.10", "Y34B4A.20"), 
                        Description = c("Affected by sir-2.1 and daf-2 according to microarray. Affected by Chlorpyrifos, dafa#1 and Sirolimus according to microarray.", "Is affected by sir-2.1 based on microarray studies.", "Affected by dpy-9, dpy-10, fbf-1, tdp-1, oga-1, ogt-1, pmk-1 and dnj-14 according to microarray. Affected by Quercetin, Hydrolyzable Tannins, paraquat, Colistin and triclosan according to microarray.", NA))
  annotations<-ungroup(annotations) %>%
  add_row(supplement)
}

# Check for presence of output folder, generate if it doesn't exist
output.path <- "./Outputs"
if (!dir.exists(output.path)){
  dir.create(output.path)
}

species <- switch(x,
                  `caelegprjna13758` = 'elegans',
                  `cabrigprjna10731` = 'briggsae',
                  `cabrenprjna20035` = 'brenneri',
                  `cajapoprjna12591` = "japonica",
                  `caremaprjna53967` = "remanei")

app.path <-"../Data"
output.name <- paste0(species, '_geneAnnotations')

save(annotations,
     file = file.path(output.path,
                      output.name))

# Save in Shiny app Data folder
save(annotations,
     file = file.path(app.path,
                      output.name))
}

```
